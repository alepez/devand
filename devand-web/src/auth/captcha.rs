use captcha::{CaptchaName, Difficulty};
use std::path::PathBuf;
use uuid::Uuid;

pub struct CaptchaFile {
    path: Box<PathBuf>,
    value: String,
}

impl CaptchaFile {
    pub fn value(&self) -> String {
        self.value.clone()
    }

    pub fn into_data(self) -> Vec<u8> {
        std::fs::read(self.path.as_ref()).unwrap()
    }
}

impl Drop for CaptchaFile {
    fn drop(&mut self) {
        std::fs::remove_file(self.path.as_ref()).unwrap();
    }
}

impl CaptchaFile {
    pub fn new() -> Self {
        let uuid = Uuid::new_v4();
        let filename = uuid.to_string() + ".png";
        let path = PathBuf::from("/tmp").join(filename);
        let path = Box::new(path);

        let c = captcha::by_name(Difficulty::Medium, CaptchaName::Mila);

        c.save(&path).expect("save failed");

        let value = c.chars_as_string();

        CaptchaFile { path, value }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn generate_and_destroy() {
        #[allow(unused_assignments)]
        let mut path = Box::new(PathBuf::default());
        {
            let x = CaptchaFile::new();
            path = x.path.clone();
            assert!(path.exists());
        }
        assert!(!path.exists());
    }

    #[test]
    fn captcha_value() {
        let x = CaptchaFile::new();
        let _value = x.value();
    }
}
