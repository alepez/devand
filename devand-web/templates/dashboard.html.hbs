{{#*inline "body"}}

    <div class="yew-mount-point"></div>

    <script src="/static/ui/devand.js"></script>

{{/inline}}
{{>layout-yew}}
