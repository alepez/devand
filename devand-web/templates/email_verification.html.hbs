{{#*inline "body"}}

{{#if verified }}
    <div class="alert alert-success">Your email has been successfully verified</div>
{{else}}
    <div class="alert alert-error">Invalid url. Verification token may be expired.</div>
{{/if}}

{{/inline}}
{{>layout}}

