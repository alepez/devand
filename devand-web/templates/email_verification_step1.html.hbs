{{#*inline "body"}}

{{#if valid_token }}
    <form action="/verify_email/{{ token }}" method="post" accept-charset="utf-8" class="pure-form pure-form-stacked">
        <div class="pure-controls">
          <button type="submit" class="pure-button pure-button-primary">Click here to complete email verification</button>
        </div>
    </form>
{{else}}
    <div class="alert alert-error">Invalid url. Verification token may be expired.</div>
{{/if}}

{{/inline}}
{{>layout}}

