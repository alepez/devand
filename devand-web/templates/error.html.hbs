{{#*inline "body"}}

<p>{{ message }}</p>

{{/inline}} {{>layout}}
