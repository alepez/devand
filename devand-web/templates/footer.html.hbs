
  <footer class="main-footer">
    <div class="main-footer-content">
      <div class="pure-menu pure-menu-horizontal">
        <ul class="pure-menu-list">
          <li class="pure-menu-item">
            <a href="https://alepez.me" class="pure-menu-link">Â© 2020 Alessandro Pezzato</a>
          </li>
          <li class="pure-menu-item">
            <a href="/privacy" class="pure-menu-link">Privacy Policy</a>
          </li>
          <li class="pure-menu-item">
            <a href="/code-of-conduct" class="pure-menu-link">Code of Conduct</a>
          </li>
          <li class="pure-menu-item">
            <a href="/help" class="pure-menu-link">Help</a>
          </li>
          <li class="pure-menu-item">
            <a href="https://github.com/alepez/devand" class="pure-menu-link">Code</a>
          </li>
        </ul>
      </div>
    </div>
  </footer>
