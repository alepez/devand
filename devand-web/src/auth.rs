mod captcha;

use self::captcha::CaptchaFile;
use crate::PgDevandConn;
use core::convert::TryFrom;
use devand_db as db;
use rocket::http::{Cookie, Cookies};
use rocket::outcome::IntoOutcome;
use rocket::request::{FromRequest, Outcome, Request};
use serde::{Deserialize, Serialize};
use validator::{Validate, ValidationError};
use validator_derive::Validate;

const LOGIN_COOKIE_KEY: &'static str = "login";
const JOIN_COOKIE_KEY: &'static str = "join";
const JOIN_CAPTCHA_COOKIE_KEY: &'static str = "join_captcha";

#[derive(FromForm)]
pub struct Credentials {
    username: String,
    password: String,
}

impl Credentials {
    fn normalize(self) -> Self {
        Self {
            username: trimlow(self.username),
            password: self.password,
        }
    }
}

impl Into<db::auth::Credentials> for Credentials {
    fn into(self) -> db::auth::Credentials {
        db::auth::Credentials {
            username: self.username,
            password: self.password,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct AuthData {
    pub user_id: i32,
}

impl AuthData {
    pub fn matches_user(&self, user: &devand_core::User) -> bool {
        user.id == self.user_id
    }
}

impl TryFrom<rocket::http::Cookie<'_>> for AuthData {
    type Error = ();
    fn try_from(cookie: rocket::http::Cookie<'_>) -> Result<Self, Self::Error> {
        let json = cookie.value();
        serde_json::from_str(json).or(Err(()))
    }
}

impl<'a> Into<rocket::http::Cookie<'a>> for AuthData {
    fn into(self) -> rocket::http::Cookie<'a> {
        let json = serde_json::to_string(&self).unwrap();
        Cookie::build(LOGIN_COOKIE_KEY, json)
            .http_only(true)
            .max_age(time::Duration::days(30))
            .finish()
    }
}

impl<'a, 'r> FromRequest<'a, 'r> for AuthData {
    type Error = ();

    fn from_request(request: &'a Request<'r>) -> Outcome<AuthData, ()> {
        request
            .cookies()
            .get_private(LOGIN_COOKIE_KEY)
            .and_then(|cookie| AuthData::try_from(cookie).ok())
            .or_forward(())
    }
}

pub(crate) fn login(
    cookies: &mut Cookies,
    credentials: Credentials,
    conn: &PgDevandConn,
) -> Result<(), ()> {
    let credentials = credentials.normalize().into();

    db::auth::login(credentials, &conn.0)
        .map(|user_id| {
            let auth_data = AuthData { user_id };
            cookies.add_private(auth_data.into());
        })
        .map_err(|_| ())
}

pub fn logout(cookies: &mut Cookies) {
    cookies.remove_private(Cookie::named(LOGIN_COOKIE_KEY));
}

#[derive(FromForm, Debug, Validate, Serialize, Deserialize, Clone)]
pub struct JoinData {
    #[validate(custom = "validate_username")]
    pub username: String,
    #[validate(email)]
    pub email: String,
    #[validate(custom = "validate_password")]
    pub password: String,
    pub captcha: String,
}

impl TryFrom<rocket::http::Cookie<'_>> for JoinData {
    type Error = ();
    fn try_from(cookie: rocket::http::Cookie<'_>) -> Result<Self, Self::Error> {
        let json = cookie.value();
        serde_json::from_str(json).or(Err(()))
    }
}

impl<'a> Into<rocket::http::Cookie<'a>> for JoinData {
    fn into(self) -> rocket::http::Cookie<'a> {
        let json = serde_json::to_string(&self).unwrap();
        Cookie::build(JOIN_COOKIE_KEY, json)
            .http_only(true)
            .max_age(time::Duration::minutes(10))
            .path("/join")
            .finish()
    }
}

impl<'a, 'r> FromRequest<'a, 'r> for JoinData {
    type Error = ();

    fn from_request(request: &'a Request<'r>) -> Outcome<JoinData, ()> {
        request
            .cookies()
            .get_private(JOIN_COOKIE_KEY)
            .and_then(|cookie| JoinData::try_from(cookie).ok())
            .or_forward(())
    }
}

impl JoinData {
    fn normalize(self) -> Self {
        Self {
            username: trimlow(self.username),
            email: trimlow(self.email),
            password: self.password,
            captcha: trimlow(self.captcha),
        }
    }
}

fn validate_password(s: &str) -> Result<(), validator::ValidationError> {
    if devand_core::auth::is_valid_password(s) {
        Ok(())
    } else {
        let mut err = ValidationError::new("invalid_password");
        err.message = Some("Make sure it's at least 15 characters OR at least 8 characters including a number and a lowercase letter".into());
        Err(err)
    }
}

fn validate_username(s: &str) -> Result<(), validator::ValidationError> {
    if devand_core::auth::is_valid_username(s) {
        Ok(())
    } else {
        let mut err = ValidationError::new("invalid_password");
        err.message = Some("Must be at least 3 characters, only lowercase and digits".into());
        Err(err)
    }
}

fn is_unique_email(s: &str, conn: &PgDevandConn) -> Result<(), validator::ValidationError> {
    if devand_db::is_email_available(s, conn) {
        Ok(())
    } else {
        let mut err = ValidationError::new("email_unavailable");
        err.message = Some("Email is already used".into());
        Err(err)
    }
}

fn is_unique_username(s: &str, conn: &PgDevandConn) -> Result<(), validator::ValidationError> {
    if devand_db::is_username_available(s, conn) {
        Ok(())
    } else {
        let mut err = ValidationError::new("username_unavailable");
        err.message = Some("Username is already used".into());
        Err(err)
    }
}

fn validate_captcha(input: String, expected: String) -> Result<(), validator::ValidationError> {
    let input = trimlow(input);
    let expected = trimlow(expected);

    if input == expected {
        Ok(())
    } else {
        let mut err = ValidationError::new("captcha");
        err.message = Some("Wrong captcha".into());
        Err(err)
    }
}

fn check_for_uniqueness(
    join_data: &JoinData,
    conn: &PgDevandConn,
) -> Result<(), validator::ValidationErrors> {
    let mut errors = validator::ValidationErrors::new();

    if let Err(e) = is_unique_username(&join_data.username, conn) {
        errors.add("username", e);
    }

    if let Err(e) = is_unique_email(&join_data.email, conn) {
        errors.add("email", e);
    }

    if errors.is_empty() {
        Ok(())
    } else {
        Err(errors)
    }
}

pub(crate) fn join(
    cookies: &mut Cookies,
    join_data: JoinData,
    expected_captcha: ExpectedCaptcha,
    conn: &PgDevandConn,
) -> Result<(), JoinError> {
    let mut join_data = join_data.normalize();

    let valid = join_data.validate();

    let captcha_input = join_data.captcha;
    join_data.captcha = String::new();

    // Remove old cookie
    cookies.remove_private(Cookie::named(JOIN_COOKIE_KEY));
    cookies.remove_private(Cookie::named(JOIN_CAPTCHA_COOKIE_KEY));

    let valid = valid
        .and_then(|_| check_for_uniqueness(&join_data, conn))
        .and_then(|_| {
            let mut errors = validator::ValidationErrors::new();

            if let Err(e) = validate_captcha(captcha_input, expected_captcha.value) {
                errors.add("captcha", e);
            }

            if errors.is_empty() {
                Ok(())
            } else {
                Err(errors)
            }
        });

    // Add new join data if there is an error
    if valid.is_err() {
        // Update with new data
        cookies.add_private(join_data.clone().into());
    }

    // Early exit if not valid
    valid?;

    let join_data = db::auth::JoinData {
        username: join_data.username,
        password: join_data.password.into(),
        email: join_data.email,
    };

    let result = db::auth::join(join_data, &conn.0);

    if result.is_ok() {
        cookies.remove_private(Cookie::named(JOIN_COOKIE_KEY));
        cookies.remove_private(Cookie::named(JOIN_CAPTCHA_COOKIE_KEY));
    }

    result.map(|_| ()).map_err(|_| JoinError::UnknownError)
}

pub(crate) fn captcha(cookies: &mut Cookies) -> Result<CaptchaFile, ()> {
    let captcha = CaptchaFile::new();

    let captcha_value = ExpectedCaptcha {
        value: captcha.value(),
    };

    cookies.add_private(captcha_value.into());

    Ok(captcha)
}

#[derive(Debug)]
pub enum JoinError {
    UnknownError,
    ValidationError(String),
}

impl From<validator::ValidationErrors> for JoinError {
    fn from(e: validator::ValidationErrors) -> Self {
        let msg = e
            .field_errors()
            .into_iter()
            .map(|(field, errors)| {
                let msg = errors
                    .iter()
                    .filter_map(|x| x.message.clone())
                    .collect::<Vec<_>>()
                    .concat();
                if msg.is_empty() {
                    format!("Invalid {}. ", field)
                } else {
                    format!("Invalid {}: {}. ", field, msg)
                }
            })
            .collect::<Vec<_>>()
            .concat();
        JoinError::ValidationError(msg)
    }
}

impl ToString for JoinError {
    fn to_string(&self) -> String {
        match self {
            JoinError::UnknownError => "UnknownError".into(),
            JoinError::ValidationError(msg) => msg.into(),
        }
    }
}

fn trimlow(s: String) -> String {
    // We help the user, trimming spaces and converting to lowercase
    let s = s.to_lowercase();
    // Note: this allocates a new string, in place trimming does not exist
    s.trim().to_string()
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ExpectedCaptcha {
    pub value: String,
}

impl TryFrom<rocket::http::Cookie<'_>> for ExpectedCaptcha {
    type Error = ();
    fn try_from(cookie: rocket::http::Cookie<'_>) -> Result<Self, Self::Error> {
        let json = cookie.value();
        serde_json::from_str(json).or(Err(()))
    }
}

impl<'a> Into<rocket::http::Cookie<'a>> for ExpectedCaptcha {
    fn into(self) -> rocket::http::Cookie<'a> {
        let json = serde_json::to_string(&self).unwrap();
        Cookie::build(JOIN_CAPTCHA_COOKIE_KEY, json)
            .http_only(true)
            .max_age(time::Duration::minutes(10))
            .path("/join")
            .finish()
    }
}

impl<'a, 'r> FromRequest<'a, 'r> for ExpectedCaptcha {
    type Error = ();

    fn from_request(request: &'a Request<'r>) -> Outcome<ExpectedCaptcha, ()> {
        request
            .cookies()
            .get_private(JOIN_CAPTCHA_COOKIE_KEY)
            .and_then(|cookie| ExpectedCaptcha::try_from(cookie).ok())
            .or_forward(())
    }
}

pub struct RealIp(pub std::net::IpAddr);

impl<'a, 'r> FromRequest<'a, 'r> for RealIp {
    type Error = ();

    fn from_request(request: &'a Request<'r>) -> Outcome<RealIp, ()> {
        let real_ip_from_header = request.real_ip();
        let remote_ip_from_sock = request.remote().map(|x| x.ip());

        real_ip_from_header
            .or(remote_ip_from_sock)
            .map(|x| RealIp(x))
            .or_forward(())
    }
}

pub struct LoggedUser(devand_core::User);

impl<'a, 'r> FromRequest<'a, 'r> for LoggedUser {
    type Error = ();

    fn from_request(request: &'a Request<'r>) -> Outcome<LoggedUser, ()> {
        let conn = request.guard::<PgDevandConn>()?;
        let auth_data = request.guard::<AuthData>()?;
        let user = devand_db::load_user_by_id(auth_data.user_id, &conn.0);
        user.map(|x| LoggedUser(x)).or_forward(())
    }
}

impl Into<devand_core::User> for LoggedUser {
    fn into(self) -> devand_core::User {
        self.0
    }
}

impl std::ops::Deref for LoggedUser {
    type Target = devand_core::User;

    #[inline(always)]
    fn deref(&self) -> &devand_core::User {
        &self.0
    }
}
