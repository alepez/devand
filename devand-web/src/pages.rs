use crate::auth::{self, AuthData, ExpectedCaptcha};
use crate::StaticDir;
use crate::{Mailer, PgDevandConn};
use devand_crypto::{EmailVerification, PasswordReset, Signable, SignedToken};
use rocket::http::{ContentType, Cookies};
use rocket::request::{FlashMessage, Form};
use rocket::response::{Content, Flash, NamedFile, Redirect};
use rocket::{Route, State};
use rocket_contrib::templates::Template;
use serde::Serialize;

const BASE_URL: Option<&'static str> = option_env!("DEVAND_BASE_URL");
const DEFAULT_BASE_URL: &'static str = "http://localhost:8000";

// Handle authentication request
#[post("/login", data = "<credentials>")]
fn login(
    mut cookies: Cookies,
    credentials: Form<auth::Credentials>,
    real_ip: auth::RealIp,
    conn: PgDevandConn,
) -> Result<Redirect, Flash<Redirect>> {
    auth::login(&mut cookies, credentials.0, &conn)
        .map(|_| Redirect::to(uri!(dashboard_index)))
        .map_err(|_| {
            log_fail(real_ip.0);
            Flash::error(
                Redirect::to(uri!(login_page)),
                "Incorrect username or password",
            )
        })
}

#[get("/login")]
fn login_page(
    auth_data: Option<auth::AuthData>,
    flash: Option<FlashMessage>,
) -> Result<Template, Redirect> {
    if auth_data.is_some() {
        // When user is authenticated, /login just redirect to index
        Err(Redirect::to(uri!(index)))
    } else {
        // When user is not authenticated, /login displays a form
        #[derive(Serialize)]
        struct Context {
            title: &'static str,
            flash_msg: Option<String>,
            flash_name: Option<String>,
            authenticated: bool,
        }

        let context = Context {
            title: "Sign in to DevAndDev",
            flash_msg: flash.as_ref().map(|x| x.msg().to_string()),
            flash_name: flash.as_ref().map(|x| x.name().to_string()),
            authenticated: false,
        };

        Ok(Template::render("login", &context))
    }
}

// TODO https://github.com/alepez/devand/issues/69 This should be POST and frontend should use a form
/// /logout just remove the cookie
#[get("/logout")]
fn logout(mut cookies: Cookies) -> Flash<Redirect> {
    auth::logout(&mut cookies);
    Flash::success(Redirect::to(uri!(login_page)), "Successfully logged out.")
}

#[get("/password_reset")]
fn password_reset_page(flash: Option<FlashMessage>) -> Template {
    #[derive(Serialize)]
    struct Context {
        title: &'static str,
        flash_msg: Option<String>,
        flash_name: Option<String>,
        authenticated: bool,
    }

    let context = Context {
        title: "Reset your password",
        flash_msg: flash.as_ref().map(|x| x.msg().to_string()),
        flash_name: flash.as_ref().map(|x| x.name().to_string()),
        authenticated: false,
    };

    Template::render("password_reset", &context)
}

#[derive(FromForm)]
pub struct PasswordReset1 {
    email: String,
}

#[post("/password_reset", data = "<password_reset>")]
fn password_reset(
    password_reset: Form<PasswordReset1>,
    real_ip: auth::RealIp,
    mailer: State<Mailer>,
    crypto_encoder: State<devand_crypto::Encoder>,
    conn: PgDevandConn,
) -> Result<Redirect, Flash<Redirect>> {
    let err_msg = "That address is not associated with a personal user account.";
    let PasswordReset1 { email } = password_reset.0;

    if let Some(user) = devand_db::load_user_by_email(email.as_str(), &conn) {
        let user_id = user.id;
        let data = PasswordReset { user_id: user_id.0 };

        let token = data.sign(&crypto_encoder);

        crate::notifications::password_reset(
            BASE_URL.unwrap_or(DEFAULT_BASE_URL),
            &mailer,
            user.email,
            token,
        );

        Ok(Redirect::to(uri!(password_reset_wait_page)))
    } else {
        log_fail(real_ip.0);
        let redirect = Redirect::to(uri!(password_reset_page));
        Err(Flash::error(redirect, err_msg))
    }
}

#[get("/password_reset/<token>")]
fn password_reset_token_page(
    token: String,
    flash: Option<FlashMessage>,
    crypto_decoder: State<devand_crypto::Decoder>,
) -> Template {
    #[derive(Serialize)]
    struct Context {
        title: &'static str,
        token: String,
        flash_msg: Option<String>,
        flash_name: Option<String>,
        authenticated: bool,
        valid_token: bool,
    }

    // Here we decode the token just to give an immediate feedback to user about its validity
    // It is checked again on form submission
    let valid_token =
        PasswordReset::try_from_token(&token.clone().into(), &crypto_decoder).is_some();

    let context = Context {
        title: "Reset your password",
        token,
        flash_msg: flash.as_ref().map(|x| x.msg().to_string()),
        flash_name: flash.as_ref().map(|x| x.name().to_string()),
        authenticated: false,
        valid_token,
    };

    Template::render("password_reset_new", &context)
}

#[get("/password_reset/wait")]
fn password_reset_wait_page() -> Template {
    #[derive(Serialize)]
    struct Context {
        title: &'static str,
        authenticated: bool,
    }

    let context = Context {
        title: "Check your email",
        authenticated: false,
    };

    Template::render("password_reset_wait", &context)
}

#[derive(FromForm)]
pub struct PasswordReset2 {
    password: String,
}

#[post("/password_reset/<token>", data = "<password_reset>")]
fn password_reset_token(
    real_ip: auth::RealIp,
    token: String,
    password_reset: Form<PasswordReset2>,
    crypto_decoder: State<devand_crypto::Decoder>,
    conn: PgDevandConn,
) -> Result<Flash<Redirect>, Flash<Redirect>> {
    let ok_msg = "Your password has been changed! You can now sign in with your new password.";
    let redirect_ok = Redirect::to(uri!(login_page));
    let redirect_err = Redirect::to(uri!(password_reset_token_page: token.clone()));

    let PasswordReset2 { password } = password_reset.0;

    if !devand_core::auth::is_valid_password(&password) {
        let err_msg = "Invalid password";
        return Err(Flash::error(redirect_err, err_msg));
    }

    let token = SignedToken::from(token);
    let signed_data = PasswordReset::try_from_token(&token, &crypto_decoder);

    match signed_data {
        Some(PasswordReset { user_id }) => {
            let user_id = devand_core::UserId(user_id);
            devand_db::auth::set_password(user_id, &password, &conn)
                .ok()
                .expect("Password to be updated on database");
            Ok(Flash::success(redirect_ok, ok_msg))
        }
        None => {
            log_fail(real_ip.0);
            let err_msg = "Invalid token";
            Err(Flash::error(redirect_err, err_msg))
        }
    }
}

// Handle join request
// Note: cookies must be after expected_captcha, due to One-At-A-Time cookies
// restriction
#[post("/join", data = "<join_data>")]
fn join(
    join_data: Form<auth::JoinData>,
    expected_captcha: ExpectedCaptcha,
    mut cookies: Cookies,
    real_ip: auth::RealIp,
    mailer: State<Mailer>,
    conn: PgDevandConn,
) -> Result<Redirect, Flash<Redirect>> {
    let email_address = join_data.email.clone();
    auth::join(&mut cookies, join_data.0, expected_captcha, &conn)
        .map(|_| mailer.verify_address(email_address))
        .map(|_| Redirect::to(uri!(dashboard_index)))
        .map_err(|err| {
            log_fail(real_ip.0);
            Flash::error(Redirect::to(uri!(join_page)), err.to_string())
        })
}

#[get("/join", rank = 2)]
fn join_page(
    flash: Option<FlashMessage>,
    join_data: Option<auth::JoinData>,
    auth_data: Option<AuthData>,
) -> Result<Template, Redirect> {
    if auth_data.is_some() {
        // When user is authenticated, /join just redirect to index
        Err(Redirect::to(uri!(index)))
    } else {
        // When user is not authenticated, /join displays a form
        #[derive(Serialize)]
        struct Context {
            title: &'static str,
            flash: Option<String>,
            username: Option<String>,
            email: Option<String>,
            password: Option<String>,
            authenticated: bool,
        }

        let context = Context {
            title: "Create your DevAndDev account",
            flash: flash.map(|x| x.msg().to_string()),
            username: join_data.as_ref().map(|x| x.username.to_string()),
            email: join_data.as_ref().map(|x| x.email.to_string()),
            password: join_data.as_ref().map(|x| x.password.to_string()),
            authenticated: false,
        };

        Ok(Template::render("join", &context))
    }
}

/// Generate a captcha png
#[get("/join/captcha.png")]
fn join_captcha(mut cookies: Cookies) -> Option<Content<Vec<u8>>> {
    let captcha = auth::captcha(&mut cookies).unwrap();
    let data = captcha.into_data();
    Some(Content(ContentType::PNG, data))
}

// Some URLs respond with dashboard
#[get("/dashboard")]
fn dashboard_index(auth_data: AuthData) -> Template {
    dashboard(auth_data)
}

#[get("/affinities")]
fn dashboard_affinities(auth_data: AuthData) -> Template {
    dashboard(auth_data)
}

#[get("/code-now")]
fn dashboard_code_now(auth_data: AuthData) -> Template {
    dashboard(auth_data)
}

#[get("/chat/<_username>")]
fn dashboard_chat(auth_data: AuthData, _username: String) -> Template {
    dashboard(auth_data)
}

#[get("/schedule")]
fn dashboard_schedule(auth_data: AuthData) -> Template {
    dashboard(auth_data)
}

#[get("/settings/password")]
fn dashboard_settings_password(auth_data: AuthData) -> Template {
    dashboard(auth_data)
}

#[get("/u/<_username>")]
fn dashboard_user_profile(auth_data: AuthData, _username: String) -> Template {
    dashboard(auth_data)
}

fn dashboard(_auth_data: AuthData) -> Template {
    #[derive(Serialize)]
    struct Context {
        title: &'static str,
        authenticated: bool,
    }

    let context = Context {
        title: "Your dashboard",
        authenticated: true,
    };

    Template::render("dashboard", &context)
}

#[get("/", rank = 2)]
fn index(auth_data: Option<AuthData>) -> Template {
    #[derive(Serialize)]
    struct Context {
        title: &'static str,
        authenticated: bool,
    }

    let context = Context {
        title: "Find your pair-programming pal",
        authenticated: auth_data.is_some(),
    };

    Template::render("index", &context)
}

#[get("/privacy")]
fn privacy(auth_data: Option<AuthData>) -> Template {
    #[derive(Serialize)]
    struct Context {
        title: &'static str,
        authenticated: bool,
    }

    let context = Context {
        title: "Privacy Policy",
        authenticated: auth_data.is_some(),
    };

    Template::render("privacy", &context)
}

#[get("/code-of-conduct")]
fn code_of_conduct(auth_data: Option<AuthData>) -> Template {
    #[derive(Serialize)]
    struct Context {
        title: &'static str,
        authenticated: bool,
    }

    let context = Context {
        title: "DevAndDev Code of Conduct",
        authenticated: auth_data.is_some(),
    };

    Template::render("code-of-conduct", &context)
}

#[get("/help")]
fn help(auth_data: Option<AuthData>) -> Template {
    #[derive(Serialize)]
    struct Context {
        title: &'static str,
        authenticated: bool,
    }

    let context = Context {
        title: "Need help?",
        authenticated: auth_data.is_some(),
    };

    Template::render("help", &context)
}

#[get("/favicon.ico")]
fn favicon(static_dir: State<StaticDir>) -> Option<NamedFile> {
    NamedFile::open(std::path::Path::new(&static_dir.0).join("favicon.ico")).ok()
}

// TODO https://github.com/alepez/devand/issues/69 This should render a form
#[get("/verify_email/<token>")]
fn verify_email_token(
    token: String,
    auth_data: Option<AuthData>,
    crypto_decoder: State<devand_crypto::Decoder>,
    conn: PgDevandConn,
) -> Template {
    let token = SignedToken::from(token);
    let verified_data = EmailVerification::try_from_token(&token, &crypto_decoder);

    let verified = if let Some(EmailVerification { address }) = verified_data {
        devand_db::set_verified_email(&address, &conn)
            .ok()
            .expect("Email can be verified");
        true
    } else {
        false
    };

    #[derive(Serialize)]
    struct Context {
        title: &'static str,
        authenticated: bool,
        verified: bool,
    }

    let context = Context {
        title: "Email address verification",
        authenticated: auth_data.is_some(),
        verified,
    };

    Template::render("email_verification", &context)
}

pub fn routes() -> Vec<Route> {
    routes![
        index,
        join,
        join_page,
        join_captcha,
        login,
        login_page,
        password_reset,
        password_reset_page,
        password_reset_token,
        password_reset_token_page,
        password_reset_wait_page,
        logout,
        dashboard_index,
        dashboard_affinities,
        dashboard_code_now,
        dashboard_schedule,
        dashboard_chat,
        dashboard_settings_password,
        dashboard_user_profile,
        privacy,
        code_of_conduct,
        help,
        favicon,
        verify_email_token,
    ]
}

fn format_fail(ip_addr: std::net::IpAddr) -> String {
    format!("Fail: IpAddr({})", ip_addr)
}

// Used by fail2ban
fn log_fail(ip_addr: std::net::IpAddr) {
    println!("{}", format_fail(ip_addr));
}

#[cfg(test)]
mod test {
    use super::*;
    use std::net::IpAddr;
    use std::net::Ipv4Addr;

    #[test]
    fn log_fail_has_expected_format() {
        assert!(format_fail(IpAddr::V4(Ipv4Addr::new(1, 2, 3, 4))) == "Fail: IpAddr(1.2.3.4)");
    }
}
