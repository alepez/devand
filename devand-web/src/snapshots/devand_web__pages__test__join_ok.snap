---
source: devand-web/src/pages.rs
expression: response.body_string().unwrap()
---

<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>DevAndDev - Create your DevAndDev account</title>
  <link rel="stylesheet" href="/static/ui-HASHPLACEHOLDER/pure.css">
  <link rel="stylesheet" href="/static/ui-HASHPLACEHOLDER/grids-responsive.css">
  <link rel="stylesheet" href="/static/ui-HASHPLACEHOLDER/main.css">
  <link rel="apple-touch-icon-precomposed" href="/static/favicon/favicon-180.png">
  <link rel="icon" href="/static/favicon/favicon-32.png" sizes="32x32">
  <link rel="icon" href="/static/favicon/favicon-128.png" sizes="128x128">
  <link rel="icon" href="/static/favicon/favicon-180.png" sizes="180x180">
  <link rel="icon" href="/static/favicon/favicon-196.png" sizes="196x196">
</head>

<body>
<div class="container">
    <header class="main-header">
    <div class="main-header-content">
      <div class="pure-menu pure-menu-horizontal">
        <ul class="pure-menu-list">
          <li class="pure-menu-item">
            <a href="/" title="DevAndDev" class="pure-menu-link"><img src="/static/favicon/favicon-128.png" /></a>
          </li>
    
          <li class="pure-menu-item">
            <a href="/login" class="pure-menu-link">Sign In</a>
          </li>
          <li class="pure-menu-item">
            <a href="/join" class="pure-menu-link">Sign Up</a>
          </li>
    
        </ul>
      </div>
    </div>
  </header>


  <div class="content">
    <h1>Create your DevAndDev account</h1>

    

    <form action="/join" method="post" accept-charset="utf-8" class="pure-form pure-form-stacked">
      <fieldset>
        <div class="pure-control-group">
          <label for="username">Username:</label>
          <input type="text" name="username" id="username" value="" placeholder="Username" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" />
          <span class="pure-form-message-inline">Choose with care, you can't change it later. At least 3 characters, only lowercase and digits.</span>
        </div>

        <div class="pure-control-group">
          <label for="email">Email:</label>
          <input type="email" inputmode="email" name="email" id="email" value="" placeholder="Email" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" />
          <span class="pure-form-message-inline">Notifications will be sent here</span>
        </div>

        <div class="pure-control-group">
          <label for="password">Password:</label>
          <input type="password" name="password" id="password" value="" placeholder="Password" />
          <span class="pure-form-message-inline">Make sure it's at least 15 characters OR at least 8 characters including a number and a lowercase letter.</span>
        </div>

        <div class="pure-control-group">
          <label for="captcha">Type the word in the image (case insensitive):</label>
          <input type="text" name="captcha" id="captcha" value="" placeholder="" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" />
          <div><img src="/join/captcha.png" /></div>
          <span class="pure-form-message-inline">As a protection against automated spam, you need to type in the word that appear in this image to register an account.</span>
        </div>

        <div class="pure-controls">
          <button type="submit" class="pure-button pure-button-primary">Create account</button>
          <span class="pure-form-message">
            By creating an account, you agree to our <a href="/code-of-conduct">Code of Conduct</a> and <a href="/privacy">Privacy Policy</a>.
          </span>
          <span class="pure-form-message">
            We’ll occasionally send you account related emails.
          </span>
        </div>
      </fieldset>
    </form>


  </div>
  <footer class="main-footer">
  <div class="main-footer-content">
    
    <div class="devand-cookie-policy">
      <p>We use cookies in the delivery of our services. To learn about the cookies we use and information about your preferences and opt-out choices, please <a href="/privacy">click here</a>. By using our platform you agree to our use of cookies.</p>
    </div>
    
    <ul class="pure-g">
      <li class="pure-u-1 pure-u-md-1-3">
        <a href="/help">Help</a>
      </li>
      <li class="pure-u-1 pure-u-md-1-3">
        <a href="/code-of-conduct">Code of Conduct</a>
      </li>
      <li class="pure-u-1 pure-u-md-1-3">
        <a href="/privacy">Privacy Policy</a>
      </li>
      <li class="pure-u-1 pure-u-md-1-3">
        <a href="https://github.com/alepez/devand">Code</a>
      </li>
      <li class="pure-u-1 pure-u-md-1-3">
        <a href="https://alepez.me">© 2020 Alessandro Pezzato</a>
      </li>
    </ul>
  </div>
</footer>

</div>
</body>

</html>


